### PR TITLE
Groupby agg works with pd.Series.nunique, but groupby nunique fails with axis=1

### LINKED ISSUE #30253: Groupby agg works with pd.Series.nunique, but groupby nunique fails with axis=1
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({
    ('A', 'B'): [1, 2],
    ('A', 'C'): [1, 3],
    ('D', 'B'): [0, 0]
}, index=pd.bdate_range('20191212', '20191213'))

# DOESNT CRASH
df.groupby(axis=1, level=0).agg(pd.Series.nunique)

# CRASH
df.groupby(axis=1, level=0).nunique()

```
#### Problem description

The groupby nunique crashes with error "ValueError: all keys need to be the same shape", whereas the groupby agg with pd.Series.nunique does not crash

#### Expected Output

pd.DataFrame({
    'A': [1, 2],
    'D': [1, 1]
}, index=pd.bdate_range('20191212', '20191213'))

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-642.6.2.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.1
pytest: 4.4.1
pip: 18.1
setuptools: 40.8.0.post20190628
Cython: 0.27.3
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: 0.8.2
IPython: 5.1.0
sphinx: 1.8.5
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.5
feather: None
matplotlib: 2.1.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by Ynob2000**
- [x] closes #30253
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -1994,3 +1994,22 @@ def test_dup_labels_output_shape(groupby_func, idx):
...
+
+    axis_number = df._get_axis_number(axis)
+    if not axis_number:
+        df = df.T
+
+    nunique_t = df.T.groupby(level=0, axis=axis).nunique().T
+
+    nunique = df.groupby(axis=int(not axis_number), level=0).nunique()
+
+    pd.testing.assert_frame_equal(nunique_t, nunique)
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]:
```suggestion
    tm.assert_frame_equal(result, expected)
```

Note will have to update variable names as well

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -1994,3 +1994,22 @@ def test_dup_labels_output_shape(groupby_func, idx):
...
     tm.assert_index_equal(result.columns, idx)
+
+
+def test_groupby_crash_on_nunique(axis):
+    # Fix following 30253
+    df = pd.DataFrame(
+        {("A", "B"): [1, 2], ("A", "C"): [1, 3], ("D", "B"): [0, 0]},
+        index=pd.bdate_range("20191212", "20191213"),
+    )
+    df.columns.names = ["COL1", "COL2"]
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]: Can just specify `columns=` in the DataFrame constructor

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -1994,3 +1994,22 @@ def test_dup_labels_output_shape(groupby_func, idx):
...
+
+    axis_number = df._get_axis_number(axis)
+    if not axis_number:
+        df = df.T
+
+    nunique_t = df.T.groupby(level=0, axis=axis).nunique().T
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]: Can you try and simplify this test a bit? Somewhat hard to follow with all of the transposing going on. Would be preferable if you can construct with literal values

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -1994,3 +1994,22 @@ def test_dup_labels_output_shape(groupby_func, idx):
 
     assert result.shape == (1, 2)
     tm.assert_index_equal(result.columns, idx)
+
+
+def test_groupby_crash_on_nunique(axis):
+    # Fix following 30253
+    df = pd.DataFrame(
+        {("A", "B"): [1, 2], ("A", "C"): [1, 3], ("D", "B"): [0, 0]},
+        index=pd.bdate_range("20191212", "20191213"),
```
Comment (WillAyd) [inline pandas/tests/groupby/test_groupby.py]: Unless these need to be date values (I don't think so) probably easier to just leave unspecified; otherwise it distracts a bit from what the test is aiming to fix

File: pandas/core/groupby/generic.py [hunk 5]
```diff
@@ -1813,10 +1813,17 @@ def groupby_series(obj, col=None):
             # Try to consolidate with normal wrapping functions
             from pandas.core.reshape.concat import concat
 
+            axis_number = obj._get_axis_number(self.axis)
+            if axis_number:
+                obj = obj.T
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Rather than transpose I think should just specify the iterator, so something like:

```
if axis_number == 0:
    iter_func = obj.items
else:
    iter_func = obj.iterrows
```

Then change subsequent line to 

```python
results = [groupby_series(content, label) for label, content in iter_func()]
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: pandas/tests/groupby/test_groupby.py (hunk 1 above)
Comment (nlepleux) [reply pandas/tests/groupby/test_groupby.py]: Sure, will do, thanks

File: pandas/tests/groupby/test_groupby.py (hunk 3 above)
Comment (nlepleux) [reply pandas/tests/groupby/test_groupby.py]: I will rework the test, with having a literal expected value and asserting the groupby is the expected literal

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (nlepleux) [reply pandas/tests/groupby/test_groupby.py]: OK, will do

File: pandas/core/groupby/generic.py (hunk 5 above)
Comment (nlepleux) [reply pandas/core/groupby/generic.py]: I'm trying that, thanks

**Comment by nlepleux** [comment]
I now have the following issues
`    
def test_nonexistent_path(all_parsers):
        # gh-2428: pls no segfault
        # gh-14086: raise more helpful FileNotFoundError
        # GH#29233 "File foo" instead of "File b'foo'"
        parser = all_parsers
        path = "{}.csv".format(tm.rands(10))
    
        msg = f"File {path} does not exist" if parser.engine == "c" else r"\[Errno 2\]"
        with pytest.raises(FileNotFoundError, match=msg) as e:
            parser.read_csv(path)
    
            filename = e.value.filename
    
>           assert path == filename
E           AssertionError: Pattern 'File uCidvRhnud.csv does not exist' not found in "[Errno 2] File b'uCidvRhnud.csv' does not exist: b'uCidvRhnud.csv'"

pandas/tests/io/parser/test_common.py:990: AssertionError
`

I cannot understand why I got those 2 tests on error, because I cannot understand how my change could impact those tests

**Comment by nlepleux** [comment]
PR has been updated accordingly to your remarks

**Review by TomAugspurger — APPROVED** [review]

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great thanks @nlepleux !

File: pandas/core/groupby/generic.py [hunk 6]
```diff
@@ -1813,9 +1813,20 @@ def groupby_series(obj, col=None):
             # Try to consolidate with normal wrapping functions
             from pandas.core.reshape.concat import concat
 
-            results = [groupby_series(content, label) for label, content in obj.items()]
+            axis_number = obj._get_axis_number(self.axis)
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: @WillAyd this is adding quite a bit of very hard and bespoke code

pls don’t merge non trivial things like this

this is also likely non performant

pls revert

**Comment by WillAyd** [comment]
What performance are you looking for from the axis=1 case? Wonder if time is better spent optimizing that than reverting

**Comment by jreback** [comment]
i am way more concerned about the code smell this introduces
w/o much review
